Create component tutorial (#208)
Financial-Times/origami-website: _tutorials/new-component/create-a-new-component-part-5.md

In `/src/js/example` setting component configuration is handled in the constructor. The `this.options` property is assigned to the given `opts` object, which is merged with any data attributes that have a namespace `data-o-example-[option]`.

For instance the `o-table` component has a sort feature which may be disabled by either passing `{sortable: false}` to the [`o-table` `init` method](https://registry.origami.ft.com/components/o-table@8.0.11/jsdoc?brand=master) or by adding the [`data-o-table-sortable="false"`](https://registry.origami.ft.com/components/o-table@8.0.11/readme?brand=master#disable-sort) attribute to the `o-table` element.

We'll add configuration options later to demonstrate. For full details see the [JavaScript configuration](/spec/v1/javascript/#configuration) section of the Origami specification.

## Interactivity

Lets start work on making our example component interactive.

We'll start by adding a `count` property, and listen for clicks on `o-example` buttons using the [`handleEvent`](https://medium.com/@WebReflection/dom-handleevent-a-cross-platform-standard-since-year-2000-5bf17287fd38) method to increment the count property.

> Origami components use browser apis directly for [Document Object Model](https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model) manipulation. For instance [`Document.querySelector`](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector) to get an element and [`HTMLElement.innerText`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/innerText) to set an elements text content.

Reviewer: For some reason, I have a feeling more people know the acronym DOM than "Document Object Model". Shall we change it to an abbreviation tag?
```suggestion
Origami components use browser apis directly for [<abbr title="Document Object Model">DOM</abbr>](https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model) manipulation. For instance [`Document.querySelector`](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector) to get an element and [`HTMLElement.innerText`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/innerText) to set an elements text content.
```

<pre><code class="o-syntax-highlight--js">	/**
	 * Class constructor.
	 * @param {HTMLElement} [exampleEl] - The component element in the DOM

> 	 * @param {Object} [options={}] - An options object for configuring the component

Reviewer: Adding the missing parameter description and missing documentation for what is returned
```suggestion
	 * @param {Event} event - The browser event which was triggered.
	 * @returns {void}
```

	 */
	constructor (exampleEl, opts) {
		this.exampleEl = exampleEl;
		this.options = Object.assign({}, {
		}, opts || Example.getDataAttributes(exampleEl));
		// A property to store the current count.
		this.count = 0;
		// Listen to all click events on the o-example instance.
		this.exampleEl.addEventListener('click', this);
	}

	/**
	 * A method to handle event listeners.
	 * https://medium.com/@WebReflection/dom-handleevent-a-cross-platform-standard-since-year-2000-5bf17287fd38
	 * https://dom.spec.whatwg.org/#dom-eventlistener-handleevent

> 	 * @param {Event} event

Reviewer: Adding the missing parameter description and missing documentation for what is returned
```suggestion
	 * @param {Event} event - The browser event which was triggered.
	 * @returns {void}
```

<pre><code class="o-syntax-highlight--diff">
&lt;div class="o-example &#123;&#123;#theme}}o-example--&#123;&#123;theme}}&#123;&#123;/theme}}" data-o-component="o-example">
-    Hello world, I am a component named o-example!
+    Hello world, I am a component named o-example! You have clicked this lovely button &lt;span data-o-example-current-count>0&lt;/span> times.
    &lt;button class="o-example__button">count&lt;/button>
&lt;/div>
</code></pre>

Now in our JavaScript we can get any current count element and update it when our click count is incremented.

<pre><code class="o-syntax-highlight--js">handleEvent(event) {
    // When any button within the `o-example` component is clicked
    // increment the count.
    if (event.target.tagName === 'BUTTON') {
        this.count++;
        // Get all the elements within o-example with
        // the attribute data-o-example-current-count.
        const countElements = this.exampleEl.querySelectorAll('[data-o-example-current-count]');
        // For each count element found, update the count.

>         countElements.forEach(e => e.innerText = this.count);

Reviewer: ESLint was marking this line with an error because the forEach function was returning the value and forEach does not use the return value.

What do we think about changing this to a for-of loop instead of a forEach?
```suggestion
		for (const element of countElements) {
			element.innerText = this.count;
		}
```

<pre><code class="o-syntax-highlight--js">handleEvent(event) {
    // When any button within the `o-example` component is clicked
    // increment the count.
    if (event.target.tagName === 'BUTTON') {
        this.count++;
        // Get all the elements within o-example with
        // the attribute data-o-example-current-count.
        const countElements = this.exampleEl.querySelectorAll('[data-o-example-current-count]');
        // For each count element found, update the count.
        // If the count is equal to or above the high count display a message instead.
        const countText = this.count < this.options.highCount ?
            this.count :
            'lots and lots of';
        countElements.forEach(e => e.innerText = countText);
    }
}</code></pre>

<figure>
	<img alt="" src="/assets/images/tutorial-new-component/hello-world-demo-14-js.png" />
	<figcaption class="o-typography-caption">
        The `o-example` component now says it has been clicked "lots and lots of time" if the click count is 100 or greater.
	</figcaption>
</figure>

## Browser Support

### Core Experience

> Most projects which use Origami components serve a reduced "core" experience to older browsers, per the [Financial Times browser support police](https://docs.google.com/document/d/1z6kecy_o9qHYIznTmqQ-IJqre72jhfd0nVa4JMsS7Q4/). The core experience at a minimum supports key and fundamental features without JavaScript. Origami components need to maintain these standards as a minimum.

Reviewer: ```suggestion
Most projects which use Origami components serve a reduced "core" experience to older browsers, per the [Financial Times browser support policy](https://docs.google.com/document/d/1z6kecy_o9qHYIznTmqQ-IJqre72jhfd0nVa4JMsS7Q4/). The core experience at a minimum supports key and fundamental features without JavaScript. Origami components need to maintain these standards as a minimum.
```

Author: 🚓

A good component to demonstrate this is [o-table](https://registry.origami.ft.com/components/o-table@8.0.11). With JavaScript available `o-table` has client-side sortable columns. When JavaScript is unavailable client side sorting is not possible, and sort buttons are not displayed in table headings. Users without JavaScript have fewer features available but are not left with confusing sort buttons which do nothing, a kind of [graceful degradation](https://developer.mozilla.org/en-US/docs/Glossary/Graceful_degradation). `o-table` also has a responsive variant which allows the table to scroll horizontally on small devices. The scrolling table works for core experience users but is enhanced with JavaScript to include arrows for a more clear indication of when scrolling is possible, a kind of [progressive enhancement](https://developer.mozilla.org/en-US/docs/Glossary/Progressive_Enhancement).

> At present our component displays a useless button for core experience users and when JavaScript [fails for some other reason](https://gds.blog.gov.uk/2013/10/21/how-many-people-are-missing-out-on-javascript-enhancement/). We can update our `o-example` component to hide the count button for core experience users and just display the hello message we have written.

Reviewer: ```suggestion
At present our component displays a useless button for core experience users and when JavaScript [fails for some other reason](https://gds.blog.gov.uk/2013/10/21/how-many-people-are-missing-out-on-javascript-enhancement/). We can update our `o-example` component to hide the count button for core experience users and only display the message we have written.
```

So we know when our component JavaScript is initiated successfully lets add a data attribute `data-o-example-js` to our component as part of the constructor:

<pre><code class="o-syntax-highlight--diff">constructor (exampleEl, opts) {
		this.exampleEl = exampleEl;
		this.options = Object.assign({}, {
		}, opts || Example.getDataAttributes(exampleEl));
		// A property to store the current count.
		this.count = 0;
		// Listen to all click events on the o-example instance.
		this.exampleEl.addEventListener('click', this);
+		// Set a data attribute so we know the component is initiated successfully.
+		// The attribute may be used in CSS to style our component conditionally.

> +		this.rootEl.setAttribute('data-o-example-js', '');

Reviewer: ```suggestion
+		this.exampleEl.setAttribute('data-o-example-js', '');
```

	// `themes` key, default to an empty list.
	$themes: map-get($opts, 'themes');
	$themes: if($themes, $themes, ());

	.o-example {
		@include oTypographyBody();
		border: 1px solid _oExampleGet('border-color');
		background: _oExampleGet('background-color');
		padding: oSpacingByName('s4');
		margin: oSpacingByName('s1');
	}

+	.o-example__counter {
+		display: none;
+	}
+
+	.o-example[data-o-example-js] .o-example__counter {
+		// Show the counter content when the javascript is initiated
+		// and the data attribute is added to the parent element
+		display: inline;
+	}

	.o-example__button {
		@include oButtonsContent($opts: ('type': 'primary'));
	}

	// Call the `oExampleAddTheme` mixin to output css
	// for each theme if the current brand supports it.
	@each $name in $themes {
		@if _oExampleSupports($name) {
			@include oExampleAddTheme($name);
		}
	}
}</code></pre>

<figure>
	<img alt="" src="/assets/images/tutorial-new-component/hello-world-demo-15-js.png" />
	<figcaption class="o-typography-caption">
        With JavaScript unavailable our component falls back to its "Hello World" without the count feature.
	</figcaption>
</figure>

> ### PolyFills

Reviewer: ```suggestion
### Polyfills
```
